### PR TITLE
Added missing information to dbc dump.

### PR DESCRIPTION
**Pull request by juleq**
I know that you do not consider the as_ functions as API, but have them working is helpful to me. Thank you for your work here.

Update: Fixed double space for non muxed signal dumps.

### CODE REVIEW

**Comment by eerimoq** [comment]
Hello,

please implement a test and I'll review again.

**Comment by juleq** [comment]
#34 I have pulled this also in here since I was having a look at the dump functionality.

File: tests/test_database.py [hunk 1]
```diff
@@ -1588,7 +1588,42 @@ def test_multiplex_bad_multiplexer(self):
...
+
+        self.assertEqual(dumped_msg.signals[0].name, "MultiplexorSig")
+        self.assertEqual(dumped_msg.signals[0].is_multiplexer, True)
+        self.assertEqual(dumped_msg.signals[0].multiplexer_ids, None)
+        self.assertEqual(dumped_msg.signals[1].name, "MultiplexedSig")
+        self.assertEqual(dumped_msg.signals[1].is_multiplexer, False)
+        self.assertEqual(dumped_msg.signals[1].multiplexer_ids[0], 0x2A)
+        self.assertEqual(dumped_msg.signals[2].name, "UnmultiplexedSig")
+        self.assertEqual(dumped_msg.signals[2].multiplexer_ids, None)
+        self.assertEqual(dumped_msg.signals[2].is_multiplexer, False)
```
Comment (eerimoq) [inline tests/test_database.py]: I recommend testing like in this test: https://github.com/eerimoq/cantools/blob/master/tests/test_database.py#L1074-L1089

That is, read the DBC, check some data, and then compare `as_dbc_string()` to the original file contents.

File: cantools/db/formats/dbc.py [hunk 2]
```diff
@@ -374,17 +374,19 @@ def _dump_messages(database):
     for message in database.messages:
         msg = []
         fmt = 'BO_ {frame_id} {name}: {length} {senders}'
-        msg.append(fmt.format(frame_id=message.frame_id,
+        msg.append(fmt.format(frame_id=message.frame_id | (1 << 31 if message.is_extended_frame else 0),
```
Comment (eerimoq) [inline cantools/db/formats/dbc.py]: This line is over  80-90 characters long. Either add line breaks, or create a local function called `get_frame_id(message)` and put the logic in that.

File: cantools/db/formats/dbc.py [hunk 3]
```diff
@@ -374,17 +374,19 @@ def _dump_messages(database):
...
 
         for signal in message.signals[::-1]:
-            fmt = (' SG_ {name} : {start}|{length}@{byte_order}{sign}'
+            fmt = (' SG_ {name}{muxType}{muxId} : {start}|{length}@{byte_order}{sign}'
                    ' ({scale},{offset})'
                    ' [{minimum}|{maximum}] "{unit}" {receivers}')
             msg.append(fmt.format(
                 name=signal.name,
+                muxType=' M' if signal.is_multiplexer else ' m' if signal.multiplexer_ids != None else '',
+                muxId='' if signal.multiplexer_ids == None else signal.multiplexer_ids[0],
```
Comment (eerimoq) [inline cantools/db/formats/dbc.py]: Lines too long. I think I prefer creating a local function `get_mux()` or similar that returns the mux string.

File: cantools/db/formats/dbc.py [hunk 4]
```diff
@@ -445,7 +447,10 @@ def _dump_attribute_definitions(database):
                                          name=attribute[2],
                                          type_=attribute[3],
                                          choices=choices))
-
+        elif attribute[0] in [ATTRIBUTE_DEFINITION]:
```
Comment (eerimoq) [inline cantools/db/formats/dbc.py]: Why not `elif attribute[0] == ATTRIBUTE_DEFINITION:`?

File: tests/test_database.py [hunk 5]
```diff
@@ -1588,7 +1588,42 @@ def test_multiplex_bad_multiplexer(self):
...
+        self.assertEqual(attr[0][0], "BA_DEF_")
+        self.assertEqual(attr[0][1], "BusType")
+        self.assertEqual(attr[0][2], "STRING")
+
+    def test_extended_id_dump(self):
+        filename = os.path.join('tests', 'files', 'test_extended_id_dump.dbc')
+        db = cantools.db.load_file(filename)
+        dumped_db = cantools.db.load_string(db.as_dbc_string())
+        reg_id_msg = dumped_db.get_message_by_frame_id(0x100)
+        ext_id_msg = dumped_db.get_message_by_frame_id(0x1C2A2A2A)
```
Comment (eerimoq) [inline tests/test_database.py]: Change `0x1C2A2A2A` to `0x1c2a2a2a`.

File: tests/test_database.py (hunk 1 above)
Comment (juleq) [reply tests/test_database.py]: Another example: Vector seems to output two newlines after the VERSION"" token. Your test files have one. If you have also used CANdb++, they might have slightly changed the dbc output filter used in CANdb++...

Comment (eerimoq) [reply tests/test_database.py]: The goal is to output DBC-files on the same format as CANdb++. Unfortunately, I've not used that tool myself, so I don't know its DBC-file formatting. Leave your file as outputted by CANdb++, and I'll consider changing the whitespace formatting later.

**Comment by juleq** [comment]
Thank you. Please comment here if I can help out with the Vector tooling I have access to. That would be CANoe 9.x and CANalyzer 10.x. Both are accompanied by CANdb++, but I would have to look up their versions. When starting from scratch, CANdb++ offers several templates. I used the empty one. Alternatives are J1939, CANoe, etc. I think they differ in predefined attributes. Best.

**Comment by eerimoq** [comment]
Any help to improve the quality of cantools is welcome. Reference test data is always useful to achieve high quality, but I can't recall any test in particular that needs improvement or should be added.

**Pull request merged**